File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_279/java/Solution.java
package com.hi.dhl.algorithms.leetcode._279.java;

import java.util.*;

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/10/7
 *     desc  :
 * </pre>
 */
class Solution {
    public int numSquares(int n) {
        List<Integer> square = new ArrayList<>();
        int sqrt = (int) Math.sqrt(n);
        for (int i = 1; i <= sqrt; i++) {
            square.add(i * i);
        }
        int step = 0;
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> set = new HashSet<>();
        queue.offer(n);
        while (!queue.isEmpty()) {
            step += 1;
            int count = queue.size();
            for (int i = 0; i < count; i++) {
                Integer target = queue.poll();
                for (Integer value : square) {
                    if (target.equals(value)) {
                        return step;
                    } else if (value > target) {
                        break;
                    } else {
                        int sub = target - value;
                        if (set.add(sub)) {
                            queue.offer(sub);
                        }
                    }
                }
            }
        }
        return step;
    }
}

File: 00-code(源代码)/src/com/hi/dhl/algorithms/leetcode/_279/kotlin/Solution.kt
package com.hi.dhl.algorithms.leetcode._279.kotlin

import java.util.*

/**
 * <pre>
 *     author: dhl
 *     date  : 2020/10/7
 *     desc  :
 * </pre>
 */

class Solution {
    fun numSquares(n: Int): Int {
        val square = arrayListOf<Int>()
        val sqrt = Math.sqrt(n.toDouble()).toInt()
        for (i in 1..sqrt) {
            square.add(i * i)
        }

        var step = 0
        val result = hashSetOf<Int>()
        val queue = LinkedList<Int>()
        queue.offer(n)
        while (!queue.isEmpty()) {
            step += 1
            val count = queue.size
            for (i in 0 until count) {
                val target = queue.poll();
                loop@ for (item in square) {
                    when {
                        target == item -> return step
                        item > target -> break@loop;
                        else -> {
                            val sub = target - item
                            if (result.add(sub)) queue.offer(sub)

                        }
                    }

                }
            }
        }
        return step
    }
}